File: kotlin/Lib/56.MergeIntervals.kt
// 56. Merge Intervals

class LeetCode56MergeIntervals {
    fun merge(intervals: Array<IntArray>): Array<IntArray> {
        if (intervals.size < 2)
            return intervals

        val list = mutableListOf<IntArray>()

        intervals.sortBy { i -> i[0] }

        var i = 0
        while (i < intervals.size) {
            var newInterval = intervals[i].clone()

            var j = i+1
            while (j < intervals.size) {
                if (!`_IsIntersect`(newInterval, intervals[j])) {
                    break
                }
                
                newInterval[0] = minOf(newInterval[0], intervals[j][0])
                newInterval[1] = maxOf(newInterval[1], intervals[j][1])
                ++j
            }
            i = j

            list.add(newInterval)
        }

        return list.toTypedArray()
    }

    private fun _IsIntersect(interval1: IntArray, interval2: IntArray): Boolean {
        val left = maxOf(interval1[0], interval2[0])
        val right = minOf(interval1[1], interval2[1])

        return left <= right
    }
}

File: csharp/src/56_MergeIntervals.cs
/*
LeetCode: 56. Merge Intervals
*/

using System;
using System.Collections.Generic;
using System.Linq;

namespace LeetCode.Problem_56
{
	public class LeetCode56MergeIntervals
	{
		private const int LEFT = 0;
		private const int RIGHT = 1;
		public int[][] Merge(int[][] intervals)
		{
			intervals = intervals.OrderBy(interval => interval[LEFT]).ToArray();

			var list = new List<int[]>();
			list.Add(intervals[0]);
			for (int i = 1; i < intervals.Length; ++i)
			{
				if (!_IsIntersect(list.Last(), intervals[i]))
				{
					list.Add(intervals[i]);
					continue;
				}

				list.Last()[RIGHT] = Math.Max(list.Last()[RIGHT], intervals[i][RIGHT]);
			}
			
			return list.ToArray();
		}
		private bool _IsIntersect(int[] interval1, int[] interval2)
		{
			return interval1[RIGHT] >= interval2[LEFT];
		}
	}
}
